connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java

```diff
@@ -0,0 +1,222 @@
+
+    public static final String OVERVIEW_DOC =
+        "Moves or copy headers on a record into fields in that record's key/value";
+
+    private static final String OPERATION_COPY = "copy";
+    private static final String OPERATION_MOVE = "move";
+
+    private static final String FIELDS_CONFIG = "fields";
+    private static final String HEADERS_CONFIG = "headers";
+    private static final String OPERATION_CONFIG = "operation";
+    private static final String OPERATION_DEFAULT_VALUE = OPERATION_COPY;
+    private static final ConfigDef.ValidString OPERATION_VALIDATOR = ConfigDef.ValidString.in(OPERATION_COPY, OPERATION_MOVE);
+
+    public static final ConfigDef CONFIG_DEF = new ConfigDef()
+        .define(FIELDS_CONFIG, ConfigDef.Type.LIST, ConfigDef.NO_DEFAULT_VALUE,
+            new NonEmptyListValidator(), ConfigDef.Importance.MEDIUM,
+            "Comma-separated list of field names whose values are to be copied/moved to headers.")
+        .define(HEADERS_CONFIG, ConfigDef.Type.LIST, ConfigDef.NO_DEFAULT_VALUE,
+            new NonEmptyListValidator(), ConfigDef.Importance.MEDIUM,
+            "Comma-separated list of headers names to be updated with the corresponding field value, in the same order as the header names listed in the headers configuration property.")
+        .define(OPERATION_CONFIG, ConfigDef.Type.STRING, OPERATION_DEFAULT_VALUE,
+            OPERATION_VALIDATOR, ConfigDef.Importance.MEDIUM,
+            "Either move if the fields are to be moved, or copy if the fields are to be just copied and left on the record.");
+        
+    private static final String PURPOSE = "set header from field";
+
+    private List<String> fieldsConfig;
+    private List<String> headersConfig;
+    private String operationConfig;
+    private Cache<Schema, Schema> schemaUpdateCache;
+
+    @Override
+    public void configure(Map<String, ?> props) {
+        final SimpleConfig config = new SimpleConfig(CONFIG_DEF, props);
+        fieldsConfig = config.getList(FIELDS_CONFIG);
+        headersConfig = config.getList(HEADERS_CONFIG);
+        operationConfig = config.getString(OPERATION_CONFIG);
+        schemaUpdateCache = new SynchronizedCache<>(new LRUCache<Schema, Schema>(16));
+    }
+
+    @Override
+    public R apply(R record) {
+        if (isTombstoneRecord(record)) {
+            return record;
+        } else if (operatingSchema(record) == null) {
+            return applySchemaless(record);
+        } else {
+            return applyWithSchema(record);
+        }
+    }
+
+    private boolean isTombstoneRecord(R record) {
+        return record.value() == null;
+    }
+
+    private R applySchemaless(R record) {
+        final Map<String, Object> value = requireMap(operatingValue(record), PURPOSE);
+
+        final Map<String, Object> updatedValue = new HashMap<>(value);
+
+        Headers updatedHeaders = new ConnectHeaders(record.headers());
+
+        int length = Math.min(headersConfig.size(), fieldsConfig.size());
```

Comment:
I think you should validate that the fields and headers properties contain the same number of literals.

Reply:
Sorry for not getting back earlier! Should I throw some exception in case of mismatch? Or is there a proper place to do this kind of validation?